nifi-nar-bundles/nifi-cybersecurity-bundle/nifi-cybersecurity-processors/src/main/java/org/apache/nifi/processors/cybersecurity/CompareFuzzyHash.java

```diff
@@ -0,0 +1,289 @@
+        // As mentioned above, add the PropertyDescriptor HASH_ALGORITHM and ATTRIBUTE_NAME from parent class
+        descriptors.add(HASH_ALGORITHM);
+        descriptors.add(ATTRIBUTE_NAME);
+        descriptors.add(MATCH_THRESHOLD);
+        descriptors.add(MATCHING_MODE);
+        this.descriptors = Collections.unmodifiableList(descriptors);
+
+        final Set<Relationship> relationships = new HashSet<Relationship>();
+        relationships.add(REL_FOUND);
+        relationships.add(REL_NOT_FOUND);
+        relationships.add(REL_FAILURE);
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        String algorithm = context.getProperty(HASH_ALGORITHM).getValue();
+
+        String inputHash = flowFile.getAttribute(context.getProperty(ATTRIBUTE_NAME).getValue());
+
+        if (inputHash == null) {
+            getLogger().info("FlowFile {} lacks the required '{}' attribute, routing to failure.",
+                    new Object[]{flowFile, context.getProperty(ATTRIBUTE_NAME).getValue() });
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        FuzzyHashMatcher fuzzyHashMatcher = null;
+
+        switch (algorithm) {
+            case tlsh:
+                fuzzyHashMatcher = new TLSHHashMatcher(getLogger());
+                break;
+            case ssdeep:
+                fuzzyHashMatcher = new SSDeepHashMatcher(getLogger());
+                break;
+        }
+
+        if (fuzzyHashMatcher.isValidHash(inputHash) == false) {
+            // and if that is the case we log
+            logger.error("Invalid hash provided. Sending to failure");
+            //  and send to failure
+            session.transfer(flowFile, REL_FAILURE);
+            session.commit();
+            return;
+        }
+
+        File file = new File(context.getProperty(HASH_LIST_FILE).getValue());
+
+        double similarity = 0;
+        double matchThreshold = context.getProperty(MATCH_THRESHOLD).asDouble();
+
+        try {
+            Map<String, Double> matched = new ConcurrentHashMap<String, Double>();
+            FileInputStream fileInputStream = new FileInputStream(file);
+            BufferedReader reader = new BufferedReader(new InputStreamReader(fileInputStream));
+
+            // If SSdeep skip the first line (as the usual format used by other tools add a header line
```

Comment:
I think the details of the file-parsing should be delegated & encapsulated by the implementations. A method in the interface like `Map<String, Double> getMatches(File definitionFile);` would then hide the various line split delimiters, etc. If you want to keep the loop logic here in the main processor, at least implement `void prepareReader(BufferedReader reader);` (no-op in most, skip a line in SSDeep), and `String getHashToCompare(String line);` -- the general idea is to remove the need for `switch` statements to handle custom per-implementation logic from this level.